### PR TITLE
JW-423 Implemented MPLSoGRE

### PR DESCRIPTION
**Pull request by Attil**

### CODE REVIEW

File: vRouter.vcxproj.user [hunk 1]
```diff
@@ -15,7 +15,8 @@
     <ScriptName>Microsoft.DriverKit.DefaultDriverPackageInstallationClass.PerformDefaultDriverPackageInstallation</ScriptName>
     <ScriptDeviceQuery />
     <DebuggerFlavor>DbgengKernelDebugger</DebuggerFlavor>
-    <TestCertificate>..\third_party\certs\codilime.com-selfsigned-cert.pfx</TestCertificate>
```
Comment (michal-clapinski) [inline vRouter.vcxproj.user]: Hmmmmmmm. Does not seem legit.

File: windows/vr_host.c [hunk 2]
```diff
@@ -108,29 +115,40 @@ complete_received_nbl(PNET_BUFFER_LIST nbl)
 {
     ASSERT(nbl != NULL);
 
+    // Cannot complete an NBL with children
+    if (nbl->ChildRefCount != 0)
+        return;
```
Comment (michal-clapinski) [inline windows/vr_host.c]: Not an assert or something?

File: windows/vr_host.c [hunk 3]
```diff
@@ -910,8 +966,9 @@ win_pull_inner_headers(struct vr_packet *pkt,
     UNREFERENCED_PARAMETER(reason);
     UNREFERENCED_PARAMETER(tunnel_type_cb);
 
-    /* Dummy implementation */
-    return 0;
+    //TODO: Implement
+
+    return 1;
```
Comment (michal-clapinski) [inline windows/vr_host.c]: Why did you change this?

File: windows/vr_host_interface.c [hunk 4]
```diff
@@ -86,6 +86,8 @@ __win_if_tx(struct vr_interface *vif, struct vr_packet *pkt)
     PNDIS_SWITCH_FORWARDING_DETAIL_NET_BUFFER_LIST_INFO fwd = NET_BUFFER_LIST_SWITCH_FORWARDING_DETAIL(nbl);
     fwd->IsPacketDataSafe = TRUE;
 
+    NdisAdvanceNetBufferListDataStart(nbl, pkt->vp_data, TRUE, NULL);
```
Comment (michal-clapinski) [inline windows/vr_host_interface.c]: Why is that here? All alone....

File: windows/vr_host.c [hunk 5]
```diff
@@ -399,14 +451,22 @@ win_pexpand_head(struct vr_packet *pkt, unsigned int hspace)
 
     pkt->vp_head =
         (unsigned char*)MmGetSystemAddressForMdlSafe(nb->CurrentMdl, LowPagePriority | MdlMappingNoExecute) + NET_BUFFER_CURRENT_MDL_OFFSET(nb);
-    pkt->vp_data += (unsigned short)hspace;
-    pkt->vp_tail += (unsigned short)hspace;
-    pkt->vp_end += (unsigned short)hspace;
+    pkt->vp_data = (unsigned short)hspace;
+    pkt->vp_tail = (unsigned short)hspace;
+    pkt->vp_end = (unsigned short)hspace;
```
Comment (michal-clapinski) [inline windows/vr_host.c]: Why did this change from '+=' to '='?

File: windows/vr_host.c (hunk 2 above)
Comment (Attil) [inline windows/vr_host.c]: No! This is important, because it WILL be called to complete an NBL with children. We just use do not do anything in that case.

File: windows/vr_host.c (hunk 5 above)
Comment (Attil) [inline windows/vr_host.c]: Because since are extending a cloned NBL (which is recommended by MS), the retreat will create a new MDL with size of `hspace`. And data, tail and end are all equal to the size of the first NBL. Which is `hspace`.

File: windows/vr_host.c (hunk 3 above)
Comment (Attil) [inline windows/vr_host.c]: This check whether we can pull the header. If I am not mistaken, on Windows using an illegal advance will simply return an error, so we do not have to be careful about it here. We should implement later some checksum checking though.

File: windows/vr_host_interface.c (hunk 4 above)
Comment (Attil) [inline windows/vr_host_interface.c]: Because in our previous implementation we always dealt with packet starting since the beggining of an MDL (so `pkt>vp_data` was 0, so the above line wouldn't do anything). This is no longer the case, so we have to decapsulate the data as requested. Otherwise we would always send the whole received packet, which is not what we want.

File: windows/vr_host.c [hunk 6]
```diff
@@ -145,6 +163,37 @@ delete_unbound_nbl(PNET_BUFFER_LIST nbl, unsigned long flags)
...
+    PNET_BUFFER_LIST nbl = NdisAllocateCloneNetBufferList(original_nbl, SxNBLPool, NULL, 0);
+
+    if (nbl == NULL)
+        goto cleanup;
+
+    nbl->ParentNetBufferList = original_nbl;
+    original_nbl->ChildRefCount++;
+
+    if (create_forwarding_context(nbl) != NDIS_STATUS_SUCCESS)
+        goto cleanup;
```
Comment (sodar) [inline windows/vr_host.c]: In case of an error, forwarding context is not freed in `cleanup` label.

File: windows/vr_host.c [hunk 7]
```diff
@@ -281,7 +326,7 @@ win_get_packet(PNET_BUFFER_LIST nbl, struct vr_interface *vif, unsigned char fla
     ULONG packet_length = NET_BUFFER_DATA_LENGTH(nb);
     ULONG left_mdl_space = current_mdl_count - current_mdl_offset;
 
-    if (pkt->vp_win_flags & VP_WIN_CREATED) {
+    if (nbl->NdisPoolHandle == SxNBLPool && nbl->ParentNetBufferList == NULL) {
```
Comment (sodar) [inline windows/vr_host.c]: Maybe we should create some macros/functions to test whether NBLs are created, cloned or received?

File: windows/vr_host.c [hunk 8]
```diff
@@ -441,21 +501,15 @@ win_pclone(struct vr_packet *pkt)
 
     ASSERT(original_nbl != NULL);
 
-    PNET_BUFFER_LIST nbl = NdisAllocateCloneNetBufferList(original_nbl, SxNBLPool, NULL, 0);
+    PNET_BUFFER_LIST nbl = clone_nbl(original_nbl);
```
Comment (sodar) [inline windows/vr_host.c]: `ChildRefCount` of `nbl` is updated twice - first time in `clone_nbl()` and the second time in line 521-528, at the of this function.

File: windows/vr_host.c [hunk 9]
```diff
@@ -145,6 +163,37 @@ delete_unbound_nbl(PNET_BUFFER_LIST nbl, unsigned long flags)
...
+clone_nbl(PNET_BUFFER_LIST original_nbl)
+{
+    ASSERT(original_nbl != NULL);
+
+    PNET_BUFFER_LIST nbl = NdisAllocateCloneNetBufferList(original_nbl, SxNBLPool, NULL, 0);
+
+    if (nbl == NULL)
+        goto cleanup;
+
+    nbl->ParentNetBufferList = original_nbl;
```
Comment (sodar) [inline windows/vr_host.c]: Maybe we should also update `SourceHandle` here? Is it necessary?

File: windows/vrouter_mod.c [hunk 10]
```diff
@@ -1068,19 +1068,5 @@ SxExtStartCompleteNetBufferListsIngress(
...
-        else
-        {
-            NdisFreeCloneNetBufferList(NetBufferLists, 0);
-        }
-    }
-    else
-    {
-        delete_unbound_nbl(NetBufferLists, SendCompleteFlags);
-    }
+    free_nbl(NetBufferLists);
```
Comment (sodar) [inline windows/vrouter_mod.c]: Should we put some assert here for checking if `NetBufferLists->Next == NULL`? `free_nbl()` does something only for the first NBL in chain.

File: windows/vr_host.c (hunk 9 above)
Comment (Attil) [inline windows/vr_host.c]: I think `NdisAllocateCloneNetBufferList` should set it to `SxNBLPool`, but I'll check it.

File: windows/vr_host.c (hunk 8 above)
Comment (Attil) [inline windows/vr_host.c]: It's a leak!

File: windows/vrouter_mod.c (hunk 10 above)
Comment (Attil) [inline windows/vrouter_mod.c]: I can do an assert. I don't think we allow any lists with more than one elements further into the code than the Start Ingress function

File: windows/vr_host.c (hunk 9 above)
Comment (Attil) [inline windows/vr_host.c]: You are right, I mistook your comment as if it was referring to `NdisPoolHandle`

**Comment by Attil** [comment]
Merged Windows changes.

File: windows/vr_host.c [hunk 11]
```diff
@@ -212,9 +237,79 @@ clone_nbl(PNET_BUFFER_LIST original_nbl)
         original_nbl->ChildRefCount--;
     }
 
+    if (fwd_ctx) {
```
Comment (sodar) [inline windows/vr_host.c]: `free_forwarding_context()` should be called before `NdisFreeCloneNetBufferList()`.

File: windows/vr_host.c [hunk 12]
```diff
@@ -416,6 +405,79 @@ win_page_free(void *address, unsigned int size)
...
+             There should be enough room between vp_head and vp_data to add packet headers.
+    */
+    pkt->vp_data = 0;
+
+    ULONG packet_length = NET_BUFFER_DATA_LENGTH(nb);
+    ULONG left_mdl_space = current_mdl_count - current_mdl_offset;
+
+    if (IS_OWNED(nbl) && !IS_CLONE(nbl))
+        pkt->vp_tail = pkt->vp_len = 0;
+    else
```
Comment (michal-clapinski) [inline windows/vr_host.c]: Can they both be true? This sounds weird.

File: windows/vr_host.c [hunk 13]
```diff
@@ -416,6 +405,79 @@ win_page_free(void *address, unsigned int size)
...
+    pkt->vp_if = vif;
+    pkt->vp_network_h = pkt->vp_inner_network_h = 0;
+    pkt->vp_nh = NULL;
+    pkt->vp_flags = 0;
+
+    // If a problem arises concerning IP checksums, tinker with:
+    // if (skb->ip_summed == CHECKSUM_PARTIAL)
+    //	pkt->vp_flags |= VP_FLAG_CSUM_PARTIAL;
+
+    pkt->vp_ttl = 64;
```
Comment (michal-clapinski) [inline windows/vr_host.c]: I vote against magic numbers in code. Const or define could be better.

File: windows/vr_host_interface.c [hunk 14]
```diff
@@ -70,17 +70,24 @@ win_if_del_tap(struct vr_interface *vif)
...
 {
     PNET_BUFFER_LIST nbl = pkt->vp_net_buffer_list;
 
-    NDIS_SWITCH_PORT_DESTINATION newDestination = { 0 };
-
-    if ((pkt->vp_win_flags & VP_WIN_CREATED) && pkt->vp_win_data > 0) {
-        NdisAdvanceNetBufferListDataStart(nbl, pkt->vp_win_data, FALSE, NULL);
-        pkt->vp_win_data = 0;
-    }
+    NDIS_SWITCH_PORT_DESTINATION newDestination = { 0 };
```
Comment (michal-clapinski) [inline windows/vr_host_interface.c]: There are useless spaces at the end of this line 🗡

**Comment by jablonskim** [comment]
Retest this please

**Comment by jablonskim** [comment]
Retest this please

**Comment by jablonskim** [comment]
Retest this please

**Comment by jablonskim** [comment]
Retest this please
